pkg/registry/client.go

```diff
@@ -209,74 +183,134 @@ func ClientOptPlainHTTP() ClientOption {
 	}
 }
 
-// ClientOptResolver returns a function that sets the resolver setting on a client options set
-func ClientOptResolver(resolver remotes.Resolver) ClientOption {
-	return func(client *Client) {
-		client.resolver = func(ref registry.Reference) (remotes.Resolver, error) {
-			return resolver, nil
-		}
-	}
-}
-
 type (
 	// LoginOption allows specifying various settings on login
-	LoginOption func(*loginOperation)
-
-	loginOperation struct {
-		username string
-		password string
-		insecure bool
-		certFile string
-		keyFile  string
-		caFile   string
-	}
+	LoginOption func(host string, client *Client) error
 )
 
 // Login logs into a registry
 func (c *Client) Login(host string, options ...LoginOption) error {
-	operation := &loginOperation{}
 	for _, option := range options {
-		option(operation)
+		if err := option(host, c); err != nil {
+			return fmt.Errorf("configuring login option: %w", err)
+		}
 	}
-	authorizerLoginOpts := []auth.LoginOption{
-		auth.WithLoginContext(ctx(c.out, c.debug)),
-		auth.WithLoginHostname(host),
-		auth.WithLoginUsername(operation.username),
-		auth.WithLoginSecret(operation.password),
-		auth.WithLoginUserAgent(version.GetUserAgent()),
-		auth.WithLoginTLS(operation.certFile, operation.keyFile, operation.caFile),
+
+	reg, err := remote.NewRegistry(host)
+	if err != nil {
+		return err
 	}
-	if operation.insecure {
-		authorizerLoginOpts = append(authorizerLoginOpts, auth.WithLoginInsecure())
+	reg.PlainHTTP = c.plainHTTP
+	reg.Client = c.authorizer
+
+	ctx := context.Background()
+
+	cred, err := c.authorizer.Credential(ctx, host)
+	if err != nil {
+		return fmt.Errorf("fetching credentials for %q: %w", host, err)
 	}
-	if err := c.authorizer.LoginWithOpts(authorizerLoginOpts...); err != nil {
+
+	if err := reg.Ping(ctx); err != nil {
+		return fmt.Errorf("authenticating to %q: %w", host, err)
+	}
+
+	key := credentials.ServerAddressFromRegistry(host)
+	if err := c.credentialsStore.Put(ctx, key, cred); err != nil {
 		return err
 	}
+
 	fmt.Fprintln(c.out, "Login Succeeded")
 	return nil
 }
 
 // LoginOptBasicAuth returns a function that sets the username/password settings on login
 func LoginOptBasicAuth(username string, password string) LoginOption {
-	return func(operation *loginOperation) {
-		operation.username = username
-		operation.password = password
+	return func(host string, client *Client) error {
```

Comment:
It looks like the return type changes here. The breaks the public API. I see this in several places.

Reply:
I can restore the previous signature and the `loginOperation` struct reduced to the parameters `host` and `client` parameters. Not sure it is worth the effort if this can only go into Helm v4.